[Bug] [Master] Stopping a workflow does not update task status correctly

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

Sometimes it happens that after stopping the workflow, the status of the killed task is still running

<img width="1701" alt="image" src="https://user-images.githubusercontent.com/11606262/208908647-222ed3e5-5bff-420c-a141-a321d910bd6e.png">

<img width="1176" alt="image" src="https://user-images.githubusercontent.com/11606262/208908697-eecec7d5-808e-4a23-8efa-12ded2473205.png">

This is a screenshot of the task configuration：
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/11606262/208909286-7a0081e5-4c81-4da8-bd8d-4571f7b4f092.png">


### What you expected to happen

Killed task status shows as killed

### How to reproduce
I think the reason is that the status update of the workflow and task is independent. The workflow is set to the stopped state before the task is completely stopped, and then deleted from the processInstanceMap, so that the taskResponse cannot be processed.

The code must be changed to reproduce as much as possible.

Modify line 175 of `org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService` and add the following code
```java
if (null != taskResponsePersistThread) {
    if (taskResponsePersistThread.addEvent(taskResponseEvent))...    
} else {
    logger.error("task response persist thread is null");
}
```

Modify line 118 of `org.apache.dolphinscheduler.server.worker.processor.TaskKillProcessor` and add the following code
```java
private Pair<Boolean, List<String>> doKill(TaskKillRequestCommand killCommand) {
    try {
        Thread.sleep(2000);
    } catch (InterruptedException e) {
        throw new RuntimeException(e);
    }
```

The reason for this is to simulate another site to remove the instance in processInstanceMap.
```java
if (taskResponsePersistThread.eventSize() == 0) {
    if (!processInstanceMap.containsKey(taskResponsePersistThread.getProcessInstanceId())) {
        processTaskResponseMap.remove(taskResponsePersistThread.getProcessInstanceId());
        logger.info("remove process instance: {}", taskResponsePersistThread.getProcessInstanceId());
    }
    continue;
}
```
 This will cause the instance in `processTaskResponseMap` to be removed in the following code, resulting in `TaskResponsePersistThread` being empty and printing 
```
task response persist thread is null
```

**master log:**
```
[INFO] 2022-12-21 20:44:27.372 org.apache.dolphinscheduler.server.master.processor.CacheProcessor:[52] - received command : CacheExpireCommand{CacheType=PROCESS_DEFINITION, cacheKey=7960963905984}
[INFO] 2022-12-21 20:44:27.390 org.apache.dolphinscheduler.server.master.processor.CacheProcessor:[70] - cache evict, type:processDefinition, key:7960963905984
[INFO] 2022-12-21 20:44:28.363 org.apache.dolphinscheduler.server.master.runner.MasterSchedulerService:[255] - find command 8, slot:0 :
[INFO] 2022-12-21 20:44:28.363 org.apache.dolphinscheduler.server.master.runner.MasterSchedulerService:[208] - find one command: id: 8, type: START_PROCESS
[INFO] 2022-12-21 20:44:28.482 org.apache.dolphinscheduler.server.master.runner.MasterSchedulerService:[221] - handle command end, command 8 process 8 start...
[INFO] 2022-12-21 20:44:28.650 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[1370] - add task to stand by list, task name:sleep-20s, task id:0, task code:7960945419712
[INFO] 2022-12-21 20:44:28.685 org.apache.dolphinscheduler.service.process.ProcessService:[1088] - start submit task : sleep-20s, instance id:8, state: RUNNING_EXECUTION
[INFO] 2022-12-21 20:44:28.803 org.apache.dolphinscheduler.service.process.ProcessService:[1101] - end submit task to db successfully:8 sleep-20s state:SUBMITTED_SUCCESS complete, instance id:8 state: RUNNING_EXECUTION  
[INFO] 2022-12-21 20:44:28.823 org.apache.dolphinscheduler.server.master.runner.task.CommonTaskProcessor:[164] - master submit success, task id:8, task name:sleep-20s, process id:8
[INFO] 2022-12-21 20:44:28.824 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[1384] - remove task from stand by list, id: 8 name:sleep-20s
[INFO] 2022-12-21 20:44:29.043 org.apache.dolphinscheduler.server.master.processor.TaskAckProcessor:[71] - taskAckCommand : TaskExecuteAckCommand{taskInstanceId=8, startTime=Wed Dec 21 20:44:29 CST 2022, host='192.168.20.62:1234', status=1, logPath='/Users/molin/dev/github/dolphinscheduler/logs/7960963905984_2/8/8.log', executePath='/tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8', processInstanceId='0'}
[INFO] 2022-12-21 20:44:29.043 org.apache.dolphinscheduler.server.master.processor.TaskAckProcessor:[71] - taskAckCommand : TaskExecuteAckCommand{taskInstanceId=8, startTime=Wed Dec 21 20:44:29 CST 2022, host='192.168.20.62:1234', status=1, logPath='/Users/molin/dev/github/dolphinscheduler/logs/7960963905984_2/8/8.log', executePath='/tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8', processInstanceId='8'}
[ERROR] 2022-12-21 20:44:29.044 org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService:[176] - task response persist thread is null
[INFO] 2022-12-21 20:44:33.368 org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService:[226] - already exists handler process size:0
[INFO] 2022-12-21 20:44:33.703 org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService:[233] - persist events 8 succeeded.
[INFO] 2022-12-21 20:44:42.226 org.apache.dolphinscheduler.server.master.processor.StateEventProcessor:[75] - received command : State Event :key: 8-0-8-0 type: PROCESS_STATE_CHANGE executeStatus: READY_STOP task instance id: 0 process instance id: 8 context: null
[INFO] 2022-12-21 20:44:42.975 org.apache.dolphinscheduler.server.master.runner.EventExecuteService:[127] - handle process instance : 8 , events count:1
[INFO] 2022-12-21 20:44:42.980 org.apache.dolphinscheduler.server.master.runner.EventExecuteService:[130] - already exists handler process size:0
[INFO] 2022-12-21 20:44:42.982 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[310] - process event: State Event :key: 8-0-8-0 type: PROCESS_STATE_CHANGE executeStatus: READY_STOP task instance id: 0 process instance id: 8 context: null
[INFO] 2022-12-21 20:44:42.983 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[497] - process:8 state RUNNING_EXECUTION change to READY_STOP
[INFO] 2022-12-21 20:44:43.013 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[1422] - kill called on process instance id: 8, num: 1
[INFO] 2022-12-21 20:44:43.094 org.apache.dolphinscheduler.server.master.runner.task.CommonTaskProcessor:[209] - master kill taskInstance name :sleep-20s taskInstance id:8
[INFO] 2022-12-21 20:44:43.095 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[310] - process event: State Event :key: null type: PROCESS_STATE_CHANGE executeStatus: STOP task instance id: 0 process instance id: 8 context: null
[INFO] 2022-12-21 20:44:43.095 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[497] - process:8 state READY_STOP change to STOP
[INFO] 2022-12-21 20:44:43.107 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread:[1320] - work flow process instance [id: 8, name:test_task_status_error-2-20221221204428415], state change from READY_STOP to STOP, cmd type: STOP
[INFO] 2022-12-21 20:44:43.307 org.apache.dolphinscheduler.server.master.runner.EventExecuteService:[139] - process instance 8 finished.
[INFO] 2022-12-21 20:44:43.385 org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService:[222] - remove process instance: 8
[INFO] 2022-12-21 20:44:43.610 org.apache.dolphinscheduler.server.master.processor.TaskResponseProcessor:[72] - received command : TaskExecuteResponseCommand{taskInstanceId=8, processInstanceId=8, status=9, startTime=null, endTime=Wed Dec 21 20:44:43 CST 2022, host=null, logPath=null, executePath=null, processId=2407, appIds='', varPool=[]}
[ERROR] 2022-12-21 20:44:43.613 org.apache.dolphinscheduler.server.master.processor.queue.TaskResponseService:[176] - task response persist thread is null

```

**worker log:**
```
[INFO] 2022-12-21 20:44:28.882 org.apache.dolphinscheduler.server.worker.processor.TaskExecuteProcessor:[110] - received command : TaskExecuteRequestCommand{taskExecutionContext='{"taskInstanceId":8,"taskName":"sleep-20s","firstSubmitTime":"2022-12-21 20:44:28","startTime":null,"endTime":null,"taskType":"SHELL","host":null,"executePath":null,"logPath":null,"taskJson":null,"processId":0,"processDefineCode":7960963905984,"processDefineVersion":2,"appIds":null,"processInstanceId":8,"scheduleTime":null,"globalParams":null,"executorId":1,"cmdTypeIfComplement":0,"tenantCode":"molin","queue":"default","projectCode":7960944102080,"taskParams":"{\"resourceList\":[],\"localParams\":[],\"rawScript\":\"echo \\\"start\\\"\\nsleep 20\\necho \\\"end\\\"\",\"dependence\":{},\"conditionResult\":{\"successNode\":[],\"failedNode\":[]},\"waitStartTimeout\":{},\"switchResult\":{}}","envFile":null,"environmentConfig":null,"definedParams":null,"taskAppId":null,"taskTimeoutStrategy":null,"taskTimeout":2147483647,"workerGroup":"default","delayTime":0,"currentExecutionStatus":null,"resources":{},"sqlTaskExecutionContext":{"warningGroupId":0,"connectionParams":null,"udfFuncTenantCodeMap":null,"defaultFS":null},"dataxTaskExecutionContext":{"dataSourceId":0,"sourcetype":0,"sourceConnectionParams":null,"dataTargetId":0,"targetType":0,"targetConnectionParams":null},"dependenceTaskExecutionContext":null,"sqoopTaskExecutionContext":{"dataSourceId":0,"sourcetype":0,"sourceConnectionParams":null,"dataTargetId":0,"targetType":0,"targetConnectionParams":null},"varPool":null,"dryRun":0,"paramsMap":null,"procedureTaskExecutionContext":{"connectionParams":null}}'}
[INFO] 2022-12-21 20:44:28.919 org.apache.dolphinscheduler.common.utils.FileUtils:[121] - create dir success /tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8
[INFO] 2022-12-21 20:44:28.921 org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread:[128] - task instance local execute path : /tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8
[INFO] 2022-12-21 20:44:29.038 org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread:[146] - the task begins to execute. task instance id: 8
[INFO] 2022-12-21 20:44:29.075 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[83] - shell task params {"resourceList":[],"localParams":[],"rawScript":"echo \"start\"\nsleep 20\necho \"end\"","dependence":{},"conditionResult":{"successNode":[],"failedNode":[]},"waitStartTimeout":{},"switchResult":{}}
[INFO] 2022-12-21 20:44:29.098 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[137] - raw script : echo "start"
sleep 20
echo "end"
[INFO] 2022-12-21 20:44:29.099 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[138] - task execute path : /tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8
[INFO] 2022-12-21 20:44:29.100 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[86] - tenantCode user:molin, task dir:8_8
[INFO] 2022-12-21 20:44:29.100 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[91] - create command file:/tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8/8_8.command
[INFO] 2022-12-21 20:44:29.101 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[117] - command : #!/bin/sh
BASEDIR=$(cd `dirname $0`; pwd)
cd $BASEDIR
source /etc/profile
/tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8/8_8_node.sh
[INFO] 2022-12-21 20:44:29.115 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[287] - task run command: sudo -u molin sh /tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8/8_8.command
[INFO] 2022-12-21 20:44:29.117 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[178] - process start, process id is: 2407
[INFO] 2022-12-21 20:44:30.121 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[66] -  -> welcome to use bigdata scheduling system...
	start
[INFO] 2022-12-21 20:44:43.101 org.apache.dolphinscheduler.server.worker.processor.TaskKillProcessor:[96] - received kill command : TaskKillRequestCommand{taskInstanceId=8}
[INFO] 2022-12-21 20:44:43.550 org.apache.dolphinscheduler.server.worker.processor.TaskKillProcessor:[145] - process id:2407, cmd:sudo -u molin kill -9 02408 02414 02415
[INFO] 2022-12-21 20:44:43.598 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[202] - process has exited, execute path:/tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8, processId:2407 ,exitStatusCode:137 ,processWaitForStatus:true ,processExitValue:137
[INFO] 2022-12-21 20:44:43.599 org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread:[206] - task instance id : 8,task final status : KILL
[INFO] 2022-12-21 20:44:43.602 org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread:[264] - develop mode is: false
[INFO] 2022-12-21 20:44:43.608 org.apache.dolphinscheduler.server.worker.runner.TaskExecuteThread:[282] - exec local path: /tmp/dolphinscheduler/exec/process/7960944102080/7960963905984_2/8/8 cleared.
[INFO] 2022-12-21 20:44:43.617 org.apache.dolphinscheduler.server.worker.processor.TaskKillProcessor:[194] - view log host : 192.168.20.62,logPath : /Users/molin/dev/github/dolphinscheduler/logs/7960963905984_2/8/8.log
[INFO] 2022-12-21 20:44:43.617 org.apache.dolphinscheduler.service.log.LogClientService:[117] - view log path /Users/molin/dev/github/dolphinscheduler/logs/7960963905984_2/8/8.log
[INFO] 2022-12-21 20:44:43.623 org.apache.dolphinscheduler.remote.NettyRemotingClient:[390] - netty client closed
[INFO] 2022-12-21 20:44:43.630 org.apache.dolphinscheduler.service.log.LogClientService:[74] - logger client closed
[INFO] 2022-12-21 20:44:44.149 TaskLogLogger-class org.apache.dolphinscheduler.plugin.task.shell.ShellTask:[60] - FINALIZE_SESSION
```

### Anything else

_No response_

### Version

2.0.x

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/HadoopUtils.java
+++ b/dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/HadoopUtils.java
@@ -92,8 +92,9 @@ public class HadoopUtils implements Closeable {
     private FileSystem fs;
 
     private HadoopUtils() {
-        init();
-        initHdfsPath();
+        if(init()) {
+            initHdfsPath();
+        }
     }
 
     public static HadoopUtils getInstance() {
@@ -120,7 +121,7 @@ public class HadoopUtils implements Closeable {
     /**
      * init hadoop configuration
      */
-    private void init() {
+    private boolean init() {
         try {
             configuration = new HdfsConfiguration();
 
@@ -168,11 +169,13 @@ public class HadoopUtils implements Closeable {
                 configuration.set(Constants.FS_S3A_ACCESS_KEY, PropertyUtils.getString(Constants.FS_S3A_ACCESS_KEY));
                 configuration.set(Constants.FS_S3A_SECRET_KEY, PropertyUtils.getString(Constants.FS_S3A_SECRET_KEY));
                 fs = FileSystem.get(configuration);
+            } else {
+                return false;
             }
-
         } catch (Exception e) {
             logger.error(e.getMessage(), e);
         }
+        return true;
     }
 
     /**

--- a/dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.java
+++ b/dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.java
@@ -84,4 +84,6 @@ public interface TaskInstanceMapper extends BaseMapper<TaskInstance> {
     TaskInstance queryLastTaskInstance(@Param("taskCode") long taskCode, @Param("startTime") Date startTime, @Param("endTime") Date endTime);
 
     List<TaskInstance> queryLastTaskInstanceList(@Param("taskCodes") Set<Long> taskCodes, @Param("startTime") Date startTime, @Param("endTime") Date endTime);
+
+    List<TaskInstance> queryTaskInstanceListByIds(@Param("ids") Set<Integer> ids);
 }

--- a/dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.xml
+++ b/dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.xml
@@ -196,4 +196,16 @@
             and start_time <![CDATA[ >= ]]> #{startTime} and start_time <![CDATA[ <= ]]> #{endTime}
         </if>
     </select>
+    <select id="queryTaskInstanceListByIds" resultType="org.apache.dolphinscheduler.dao.entity.TaskInstance">
+        select
+        <include refid="baseSql"/>
+        from t_ds_task_instance
+        where 1=1
+        <if test="ids != null and ids.size() != 0">
+            and id in
+            <foreach collection="ids" index="index" item="i" open="(" separator="," close=")">
+                #{i}
+            </foreach>
+        </if>
+    </select>
 </mapper>

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskKillResponseProcessor.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/TaskKillResponseProcessor.java
@@ -71,10 +71,8 @@ public class TaskKillResponseProcessor implements NettyRequestProcessor {
         TaskKillResponseCommand responseCommand = JSONUtils.parseObject(command.getBody(), TaskKillResponseCommand.class);
         logger.info("received task kill response command : {}", responseCommand);
         // TaskResponseEvent
-        TaskResponseEvent taskResponseEvent = TaskResponseEvent.newActionStop(ExecutionStatus.of(responseCommand.getStatus()),
-                responseCommand.getTaskInstanceId(),
-                responseCommand.getProcessInstanceId()
-        );
+        TaskResponseEvent taskResponseEvent = TaskResponseEvent.newKillResponse(ExecutionStatus.of(responseCommand.getStatus()),
+            responseCommand.getTaskInstanceId(), channel, responseCommand.getProcessInstanceId());
         taskResponseService.addResponse(taskResponseEvent);
     }
 

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponseEvent.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponseEvent.java
@@ -100,6 +100,19 @@ public class TaskResponseEvent {
      */
     private long opaque;
 
+    public static TaskResponseEvent newKillResponse(ExecutionStatus state,
+                                                    int taskInstanceId,
+                                                    Channel channel,
+                                                    int processInstanceId) {
+        TaskResponseEvent event = new TaskResponseEvent();
+        event.setState(state);
+        event.setTaskInstanceId(taskInstanceId);
+        event.setEvent(Event.ACTION_STOP);
+        event.setChannel(channel);
+        event.setProcessInstanceId(processInstanceId);
+        return event;
+    }
+
     public static TaskResponseEvent newActionStop(ExecutionStatus state,
                                                   int taskInstanceId,
                                                   int processInstanceId) {

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponsePersistThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponsePersistThread.java
@@ -26,7 +26,6 @@ import org.apache.dolphinscheduler.remote.command.DBTaskAckCommand;
 import org.apache.dolphinscheduler.remote.command.DBTaskResponseCommand;
 import org.apache.dolphinscheduler.remote.command.TaskKillAckCommand;
 import org.apache.dolphinscheduler.remote.command.TaskRecallAckCommand;
-import org.apache.dolphinscheduler.remote.processor.NettyRemoteChannel;
 import org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThread;
 import org.apache.dolphinscheduler.server.master.runner.task.ITaskProcessor;
 import org.apache.dolphinscheduler.server.master.runner.task.TaskAction;
@@ -159,6 +158,7 @@ public class TaskResponsePersistThread implements Runnable {
                         taskProcessor.persist(TaskAction.STOP);
                         logger.debug("ACTION_STOP: task instance id:{}, process instance id:{}", taskResponseEvent.getTaskInstanceId(), taskResponseEvent.getProcessInstanceId());
                     }
+                    workflowExecuteThread.getActiveTaskProcessorMaps().remove(taskResponseEvent.getTaskInstanceId());
                 }
 
                 if (channel != null) {
@@ -197,7 +197,7 @@ public class TaskResponsePersistThread implements Runnable {
         }
 
         WorkflowExecuteThread workflowExecuteThread = this.processInstanceMapper.get(taskResponseEvent.getProcessInstanceId());
-        if (workflowExecuteThread != null && taskResponseEvent.getState().typeIsFinished()) {
+        if (workflowExecuteThread != null && taskResponseEvent.getState().typeIsFinished() && event != Event.ACTION_STOP) {
             StateEvent stateEvent = new StateEvent();
             stateEvent.setProcessInstanceId(taskResponseEvent.getProcessInstanceId());
             stateEvent.setTaskInstanceId(taskResponseEvent.getTaskInstanceId());

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponsePersistThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponsePersistThread.java
@@ -159,6 +159,9 @@ public class TaskResponsePersistThread implements Runnable {
                         logger.debug("ACTION_STOP: task instance id:{}, process instance id:{}", taskResponseEvent.getTaskInstanceId(), taskResponseEvent.getProcessInstanceId());
                     }
                     workflowExecuteThread.getActiveTaskProcessorMaps().remove(taskResponseEvent.getTaskInstanceId());
+                    if (workflowExecuteThread.activeTaskFinish()) {
+                        this.processInstanceMapper.remove(taskResponseEvent.getProcessInstanceId());
+                    }
                 }
 
                 if (channel != null) {
@@ -197,7 +200,8 @@ public class TaskResponsePersistThread implements Runnable {
         }
 
         WorkflowExecuteThread workflowExecuteThread = this.processInstanceMapper.get(taskResponseEvent.getProcessInstanceId());
-        if (workflowExecuteThread != null && taskResponseEvent.getState().typeIsFinished() && event != Event.ACTION_STOP) {
+        if (workflowExecuteThread != null && taskResponseEvent.getState().typeIsFinished()
+            && event != Event.ACTION_STOP && !workflowExecuteThread.getProcessInstance().getState().typeIsStop()) {
             StateEvent stateEvent = new StateEvent();
             stateEvent.setProcessInstanceId(taskResponseEvent.getProcessInstanceId());
             stateEvent.setTaskInstanceId(taskResponseEvent.getTaskInstanceId());

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponseService.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/processor/queue/TaskResponseService.java
@@ -177,7 +177,7 @@ public class TaskResponseService {
                     Thread.currentThread().interrupt();
                     break;
                 } catch (Exception e) {
-                    logger.error("persist task error", e);
+                    logger.error("handle task error", e);
                 }
             }
             logger.info("StateEventResponseWorker stopped");
@@ -227,7 +227,7 @@ public class TaskResponseService {
                 FutureCallback futureCallback = new FutureCallback() {
                     @Override
                     public void onSuccess(Object o) {
-                        logger.info("persist events {} succeeded.", taskResponsePersistThread.getProcessInstanceId());
+                        logger.info("handle events {} succeeded.", taskResponsePersistThread.getProcessInstanceId());
                         if (!processInstanceMap.containsKey(taskResponsePersistThread.getProcessInstanceId())) {
                             processTaskResponseMap.remove(taskResponsePersistThread.getProcessInstanceId());
                             logger.info("remove process instance: {}", taskResponsePersistThread.getProcessInstanceId());
@@ -237,7 +237,7 @@ public class TaskResponseService {
 
                     @Override
                     public void onFailure(Throwable throwable) {
-                        logger.error("persist events failed: {}", throwable);
+                        logger.error("handle events failed: {}", throwable.getMessage());
                         if (!processInstanceMap.containsKey(taskResponsePersistThread.getProcessInstanceId())) {
                             processTaskResponseMap.remove(taskResponsePersistThread.getProcessInstanceId());
                             logger.info("remove process instance: {}", taskResponsePersistThread.getProcessInstanceId());

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/EventExecuteService.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/EventExecuteService.java
@@ -133,7 +133,7 @@ public class EventExecuteService extends Thread {
             FutureCallback futureCallback = new FutureCallback() {
                 @Override
                 public void onSuccess(Object o) {
-                    if (workflowExecuteThread.workFlowFinish()) {
+                    if (workflowExecuteThread.workFlowFinish() && workflowExecuteThread.activeTaskFinish()) {
                         processInstanceExecMaps.remove(processInstanceId);
                         notifyProcessChanged();
                         logger.info("process instance {} finished.", processInstanceId);

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
@@ -942,11 +942,6 @@ public class WorkflowExecuteThread implements Runnable {
                 taskInstance.setVarPool(JSONUtils.toJsonString(allProperty.values()));
             }
         }
-//        else {
-//            if (StringUtils.isNotEmpty(processInstance.getVarPool())) {
-//                taskInstance.setVarPool(processInstance.getVarPool());
-//            }
-//        }
     }
 
     private void setVarPoolValue(Map<String, Property> allProperty, Map<String, TaskInstance> allTaskInstance, TaskInstance preTaskInstance, Property thisProperty) {
@@ -1455,6 +1450,19 @@ public class WorkflowExecuteThread implements Runnable {
 
     public boolean workFlowFinish() {
         return this.processInstance.getState().typeIsFinished();
+    }
+
+    public boolean activeTaskFinish() {
+        if (activeTaskProcessorMaps.isEmpty()) {
+            return true;
+        }
+        List<TaskInstance> taskInstanceList = processService.findTaskInstanceListByIds(activeTaskProcessorMaps.keySet());
+        for (TaskInstance taskInstance : taskInstanceList) {
+            if (!taskInstance.getState().typeIsFinished()) {
+                return false;
+            }
+        }
+        return true;
     }
 
     /**

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
@@ -1518,6 +1518,7 @@ public class WorkflowExecuteThread implements Runnable {
             for (TaskInstance task : skipSubmitInstance) {
                 readyToSubmitTaskQueue.put(task);
             }
+            skipSubmitInstance.clear();
         } catch (Exception e) {
             logger.error("submit standby task error", e);
         }

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
@@ -1452,6 +1452,7 @@ public class WorkflowExecuteThread implements Runnable {
     private void submitStandByTask() {
         try {
             int length = readyToSubmitTaskQueue.size();
+            List<TaskInstance> skipSubmitInstance = new ArrayList<>();
             for (int i = 0; i < length; i++) {
                 TaskInstance task = readyToSubmitTaskQueue.peek();
                 if (task == null) {
@@ -1472,6 +1473,8 @@ public class WorkflowExecuteThread implements Runnable {
                         long failedTimeInterval = DateUtils.differSec(new Date(), retryTask.getEndTime());
                         if ((long) retryTask.getRetryInterval() * SEC_2_MINUTES_TIME_UNIT > failedTimeInterval) {
                             logger.info("task name: {} retry waiting has not exceeded the interval time, and skip submission this time, task id:{}", task.getName(), task.getId());
+                            readyToSubmitTaskQueue.remove(task);
+                            skipSubmitInstance.add(task);
                             continue;
                         }
                     }
@@ -1511,6 +1514,9 @@ public class WorkflowExecuteThread implements Runnable {
                     removeTaskFromStandbyList(task);
                     logger.info("remove task {},id:{} , because depend result : {}", task.getName(), task.getId(), dependResult);
                 }
+            }
+            for (TaskInstance task : skipSubmitInstance) {
+                readyToSubmitTaskQueue.put(task);
             }
         } catch (Exception e) {
             logger.error("submit standby task error", e);

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java
@@ -1452,7 +1452,7 @@ public class WorkflowExecuteThread implements Runnable {
     private void submitStandByTask() {
         try {
             int length = readyToSubmitTaskQueue.size();
-            List<TaskInstance> skipSubmitInstance = new ArrayList<>();
+            List<TaskInstance> skipSubmitInstances = new ArrayList<>();
             for (int i = 0; i < length; i++) {
                 TaskInstance task = readyToSubmitTaskQueue.peek();
                 if (task == null) {
@@ -1474,7 +1474,7 @@ public class WorkflowExecuteThread implements Runnable {
                         if ((long) retryTask.getRetryInterval() * SEC_2_MINUTES_TIME_UNIT > failedTimeInterval) {
                             logger.info("task name: {} retry waiting has not exceeded the interval time, and skip submission this time, task id:{}", task.getName(), task.getId());
                             readyToSubmitTaskQueue.remove(task);
-                            skipSubmitInstance.add(task);
+                            skipSubmitInstances.add(task);
                             continue;
                         }
                     }
@@ -1515,10 +1515,10 @@ public class WorkflowExecuteThread implements Runnable {
                     logger.info("remove task {},id:{} , because depend result : {}", task.getName(), task.getId(), dependResult);
                 }
             }
-            for (TaskInstance task : skipSubmitInstance) {
+            for (TaskInstance task : skipSubmitInstances) {
                 readyToSubmitTaskQueue.put(task);
             }
-            skipSubmitInstance.clear();
+            skipSubmitInstances.clear();
         } catch (Exception e) {
             logger.error("submit standby task error", e);
         }

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskExecuteProcessor.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskExecuteProcessor.java
@@ -88,8 +88,6 @@ public class TaskExecuteProcessor implements NettyRequestProcessor {
      * @param taskExecutionContext task
      */
     private void setTaskCache(TaskExecutionContext taskExecutionContext) {
-        TaskExecutionContext preTaskCache = new TaskExecutionContext();
-        preTaskCache.setTaskInstanceId(taskExecutionContext.getTaskInstanceId());
         TaskRequest taskRequest = JSONUtils.parseObject(JSONUtils.toJsonString(taskExecutionContext), TaskRequest.class);
         TaskExecutionContextCacheManager.cacheTaskExecutionContext(taskRequest);
     }

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskKillAckProcessor.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskKillAckProcessor.java
@@ -41,20 +41,19 @@ public class TaskKillAckProcessor implements NettyRequestProcessor {
     public void process(Channel channel, Command command) {
         Preconditions.checkArgument(CommandType.TASK_KILL_RESPONSE_ACK == command.getType(),
                 String.format("invalid command type : %s", command.getType()));
-
-        TaskKillAckCommand taskKillAckCommand = JSONUtils.parseObject(
-                command.getBody(), TaskKillAckCommand.class);
-
+        TaskKillAckCommand taskKillAckCommand = JSONUtils.parseObject(command.getBody(), TaskKillAckCommand.class);
         if (taskKillAckCommand == null) {
+            logger.warn("Cannot parse command, command type: {}", command.getType());
             return;
         }
+        logger.info("received kill ack command : {}", taskKillAckCommand);
 
         if (taskKillAckCommand.getStatus() == ExecutionStatus.SUCCESS.getCode()) {
             ResponceCache.get().removeKillResponseCache(taskKillAckCommand.getTaskInstanceId());
             TaskExecutionContextCacheManager.removeByTaskInstanceId(taskKillAckCommand.getTaskInstanceId());
-            logger.debug("removeKillResponseCache: task instance id:{}", taskKillAckCommand.getTaskInstanceId());
+            logger.info("removeKillResponseCache: task instance id:{}", taskKillAckCommand.getTaskInstanceId());
             TaskCallbackService.remove(taskKillAckCommand.getTaskInstanceId());
-            logger.debug("remove REMOTE_CHANNELS, task instance id:{}", taskKillAckCommand.getTaskInstanceId());
+            logger.info("remove REMOTE_CHANNELS, task instance id:{}", taskKillAckCommand.getTaskInstanceId());
         }
     }
 }

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskKillProcessor.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskKillProcessor.java
@@ -95,16 +95,18 @@ public class TaskKillProcessor implements NettyRequestProcessor {
         TaskKillRequestCommand killCommand = JSONUtils.parseObject(command.getBody(), TaskKillRequestCommand.class);
         logger.info("received kill command : {}", killCommand);
 
-        taskCallbackService.addRemoteChannel(killCommand.getTaskInstanceId(),
-                new NettyRemoteChannel(channel, command.getOpaque()));
+        TaskRequest taskRequest = TaskExecutionContextCacheManager.getByTaskInstanceId(killCommand.getTaskInstanceId());
+        if (taskRequest == null) {
+            logger.warn("Cannot find taskInstanceId {} in taskContextCacheManager", killCommand.getTaskInstanceId());
+            return;
+        }
+        taskRequest.setCurrentExecutionStatus(org.apache.dolphinscheduler.spi.task.ExecutionStatus.STOP);
+        TaskExecutionContextCacheManager.updateTaskExecutionContext(taskRequest);
+
+        taskCallbackService.addRemoteChannel(killCommand.getTaskInstanceId(), new NettyRemoteChannel(channel, command.getOpaque()));
 
         Pair<Boolean, List<String>> result = doKill(killCommand);
 
-        TaskRequest taskRequest = TaskExecutionContextCacheManager.getByTaskInstanceId(killCommand.getTaskInstanceId());
-
-        if (taskRequest == null) {
-            return;
-        }
         TaskKillResponseCommand taskKillResponseCommand = buildKillTaskResponseCommand(taskRequest, result);
         ResponceCache.get().cache(taskKillResponseCommand.getTaskInstanceId(), taskKillResponseCommand.convert2Command(), Event.ACTION_STOP);
         taskCallbackService.sendResult(taskKillResponseCommand.getTaskInstanceId(), taskKillResponseCommand.convert2Command());

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java
@@ -141,11 +141,11 @@ public class TaskExecuteThread implements Runnable, Delayed {
                 taskExecutionContext.setStartTime(new Date());
             }
             if (taskExecutionContext.getCurrentExecutionStatus() != ExecutionStatus.RUNNING_EXECUTION) {
-                changeTaskExecutionStatusToRunning();
+                //changeTaskExecutionStatusToRunning();
+                logger.info("the task begins to execute. task instance id: {}", taskExecutionContext.getTaskInstanceId());
+                taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.RUNNING_EXECUTION);
+                sendTaskExecuteRunningCommand(taskExecutionContext);
             }
-            logger.info("the task begins to execute. task instance id: {}", taskExecutionContext.getTaskInstanceId());
-            taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.RUNNING_EXECUTION);
-            sendTaskExecuteRunningCommand(taskExecutionContext);
             int dryRun = taskExecutionContext.getDryRun();
             // copy hdfs/minio file to local
             if (dryRun == Constants.DRY_RUN_FLAG_NO) {
@@ -212,9 +212,15 @@ public class TaskExecuteThread implements Runnable, Delayed {
             responseCommand.setProcessId(task.getProcessId());
             responseCommand.setAppIds(task.getAppIds());
         } finally {
-            TaskExecutionContextCacheManager.removeByTaskInstanceId(taskExecutionContext.getTaskInstanceId());
-            ResponceCache.get().cache(taskExecutionContext.getTaskInstanceId(), responseCommand.convert2Command(), Event.RESULT);
-            taskCallbackService.sendResult(taskExecutionContext.getTaskInstanceId(), responseCommand.convert2Command());
+            if (this.task.getExitStatusCode() == TaskConstants.EXIT_CODE_KILL
+                && TaskExecutionContextCacheManager.findTaskExecutionStatus(taskExecutionContext.getTaskInstanceId()).typeIsStop()) {
+                logger.info("task has exited, taskInstanceId:{}, exitStatusCode:{} ,task executionStatus:{}",
+                    taskExecutionContext.getTaskInstanceId(), this.task.getExitStatusCode(), ExecutionStatus.STOP);
+            } else {
+                TaskExecutionContextCacheManager.removeByTaskInstanceId(taskExecutionContext.getTaskInstanceId());
+                ResponceCache.get().cache(taskExecutionContext.getTaskInstanceId(), responseCommand.convert2Command(), Event.RESULT);
+                taskCallbackService.sendResult(taskExecutionContext.getTaskInstanceId(), responseCommand.convert2Command());
+            }
             clearTaskExecPath();
         }
     }
@@ -350,42 +356,6 @@ public class TaskExecuteThread implements Runnable, Delayed {
                 logger.info("file : {} exists ", resFile.getName());
             }
         }
-    }
-
-    /**
-     * send an ack to change the status of the task.
-     */
-    private void changeTaskExecutionStatusToRunning() {
-        taskExecutionContext.setCurrentExecutionStatus(ExecutionStatus.RUNNING_EXECUTION);
-        Command ackCommand = buildAckCommand().convert2Command();
-        try {
-            RetryerUtils.retryCall(() -> {
-                taskCallbackService.sendAck(taskExecutionContext.getTaskInstanceId(), ackCommand);
-                return Boolean.TRUE;
-            });
-        } catch (ExecutionException | RetryException e) {
-            logger.error(e.getMessage(), e);
-        }
-    }
-
-    /**
-     * build ack command.
-     *
-     * @return TaskExecuteAckCommand
-     */
-    private TaskExecuteAckCommand buildAckCommand() {
-        TaskExecuteAckCommand ackCommand = new TaskExecuteAckCommand();
-        ackCommand.setTaskInstanceId(taskExecutionContext.getTaskInstanceId());
-        ackCommand.setStatus(taskExecutionContext.getCurrentExecutionStatus().getCode());
-        ackCommand.setStartTime(taskExecutionContext.getStartTime());
-        ackCommand.setLogPath(taskExecutionContext.getLogPath());
-        ackCommand.setHost(taskExecutionContext.getHost());
-        if (TaskType.SQL.getDesc().equalsIgnoreCase(taskExecutionContext.getTaskType()) || TaskType.PROCEDURE.getDesc().equalsIgnoreCase(taskExecutionContext.getTaskType())) {
-            ackCommand.setExecutePath(null);
-        } else {
-            ackCommand.setExecutePath(taskExecutionContext.getExecutePath());
-        }
-        return ackCommand;
     }
 
     /**

--- a/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java
+++ b/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java
@@ -20,10 +20,14 @@ package org.apache.dolphinscheduler.server.worker.runner;
 import org.apache.dolphinscheduler.common.Constants;
 import org.apache.dolphinscheduler.common.enums.Event;
 import org.apache.dolphinscheduler.common.enums.ExecutionStatus;
-import org.apache.dolphinscheduler.common.enums.TaskType;
 import org.apache.dolphinscheduler.common.process.Property;
-import org.apache.dolphinscheduler.common.utils.*;
-import org.apache.dolphinscheduler.remote.command.Command;
+import org.apache.dolphinscheduler.common.utils.CommonUtils;
+import org.apache.dolphinscheduler.common.utils.DateUtils;
+import org.apache.dolphinscheduler.common.utils.FileUtils;
+import org.apache.dolphinscheduler.common.utils.HadoopUtils;
+import org.apache.dolphinscheduler.common.utils.JSONUtils;
+import org.apache.dolphinscheduler.common.utils.LoggerUtils;
+import org.apache.dolphinscheduler.common.utils.OSUtils;
 import org.apache.dolphinscheduler.remote.command.TaskExecuteAckCommand;
 import org.apache.dolphinscheduler.remote.command.TaskExecuteResponseCommand;
 import org.apache.dolphinscheduler.server.utils.LogUtils;
@@ -51,14 +55,11 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Delayed;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import com.github.rholder.retry.RetryException;
 
 /**
  * task scheduler thread
@@ -168,13 +169,19 @@ public class TaskExecuteThread implements Runnable, Delayed {
                 throw new RuntimeException(String.format("%s Task Plugin Not Found,Please Check Config File.", taskExecutionContext.getTaskType()));
             }
             TaskRequest taskRequest = JSONUtils.parseObject(JSONUtils.toJsonString(taskExecutionContext), TaskRequest.class);
+            if (null == taskRequest) {
+                throw new RuntimeException("The taskExecutionContext parse error");
+            }
             String taskLogName = LoggerUtils.buildTaskId(LoggerUtils.TASK_LOGGER_INFO_PREFIX,
                     taskExecutionContext.getProcessDefineCode(),
                     taskExecutionContext.getProcessDefineVersion(),
                     taskExecutionContext.getProcessInstanceId(),
                     taskExecutionContext.getTaskInstanceId());
             taskRequest.setTaskLogName(taskLogName);
-
+            if (!TaskExecutionContextCacheManager.updateTaskExecutionContext(taskRequest)) {
+                TaskExecutionContextCacheManager.cacheTaskExecutionContext(taskRequest);
+                logger.info("taskRequest reCache successfully, taskInstanceId: {}", taskExecutionContext.getTaskInstanceId());
+            }
             // set the name of the current thread
             Thread.currentThread().setName(String.format(TaskConstants.TASK_LOGGER_THREAD_NAME_FORMAT,taskLogName));
 
@@ -212,9 +219,8 @@ public class TaskExecuteThread implements Runnable, Delayed {
             responseCommand.setProcessId(task.getProcessId());
             responseCommand.setAppIds(task.getAppIds());
         } finally {
-            if (this.task.getExitStatusCode() == TaskConstants.EXIT_CODE_KILL
-                && TaskExecutionContextCacheManager.findTaskExecutionStatus(taskExecutionContext.getTaskInstanceId()).typeIsStop()) {
-                logger.info("task has exited, taskInstanceId:{}, exitStatusCode:{} ,task executionStatus:{}",
+            if (TaskExecutionContextCacheManager.statusIsStop(taskExecutionContext.getTaskInstanceId())) {
+                logger.info("task has exited, taskInstanceId:{}, exitStatusCode:{}, task executionStatus:{}",
                     taskExecutionContext.getTaskInstanceId(), this.task.getExitStatusCode(), ExecutionStatus.STOP);
             } else {
                 TaskExecutionContextCacheManager.removeByTaskInstanceId(taskExecutionContext.getTaskInstanceId());

--- a/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java
+++ b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java
@@ -1489,10 +1489,23 @@ public class ProcessService {
      * find task instance by id
      *
      * @param taskId task id
-     * @return task intance
+     * @return task instance
      */
     public TaskInstance findTaskInstanceById(Integer taskId) {
         return taskInstanceMapper.selectById(taskId);
+    }
+
+    /**
+     * find task instance list by ids
+     *
+     * @param taskIds task id list
+     * @return task instance list
+     */
+    public List<TaskInstance> findTaskInstanceListByIds(Set<Integer> taskIds) {
+        if (CollectionUtils.isEmpty(taskIds)) {
+            return new ArrayList<>();
+        }
+        return taskInstanceMapper.queryTaskInstanceListByIds(taskIds);
     }
 
     /**

--- a/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/TaskExecutionContextCacheManager.java
+++ b/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/TaskExecutionContextCacheManager.java
@@ -72,7 +72,11 @@ public class TaskExecutionContextCacheManager {
         return taskRequestContextCache.values();
     }
 
-    public static ExecutionStatus findTaskExecutionStatus(Integer taskInstanceId) {
-        return taskRequestContextCache.get(taskInstanceId).getCurrentExecutionStatus();
+    public static boolean statusIsStop(Integer taskInstanceId) {
+        TaskRequest taskRequest = taskRequestContextCache.get(taskInstanceId);
+        if (taskRequest == null) {
+            return true;
+        }
+        return taskRequest.getCurrentExecutionStatus().typeIsStop();
     }
 }

--- a/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/TaskExecutionContextCacheManager.java
+++ b/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/TaskExecutionContextCacheManager.java
@@ -71,4 +71,8 @@ public class TaskExecutionContextCacheManager {
     public static Collection<TaskRequest> getAllTaskRequestList() {
         return taskRequestContextCache.values();
     }
+
+    public static ExecutionStatus findTaskExecutionStatus(Integer taskInstanceId) {
+        return taskRequestContextCache.get(taskInstanceId).getCurrentExecutionStatus();
+    }
 }

--- a/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/request/TaskRequest.java
+++ b/dolphinscheduler-spi/src/main/java/org/apache/dolphinscheduler/spi/task/request/TaskRequest.java
@@ -18,6 +18,7 @@
 package org.apache.dolphinscheduler.spi.task.request;
 
 import org.apache.dolphinscheduler.spi.enums.TaskTimeoutStrategy;
+import org.apache.dolphinscheduler.spi.task.ExecutionStatus;
 import org.apache.dolphinscheduler.spi.task.Property;
 
 import java.util.Date;
@@ -182,6 +183,11 @@ public class TaskRequest {
      * delay execution time.
      */
     private int delayTime;
+
+    /**
+     * current execution status
+     */
+    private ExecutionStatus currentExecutionStatus;
 
     /**
      *  Task Logger name should be like: Task-{processDefinitionId}-{processInstanceId}-{taskInstanceId}
@@ -471,6 +477,14 @@ public class TaskRequest {
 
     public void setDelayTime(int delayTime) {
         this.delayTime = delayTime;
+    }
+
+    public ExecutionStatus getCurrentExecutionStatus() {
+        return currentExecutionStatus;
+    }
+
+    public void setCurrentExecutionStatus(ExecutionStatus currentExecutionStatus) {
+        this.currentExecutionStatus = currentExecutionStatus;
     }
 
     public SQLTaskExecutionContext getSqlTaskExecutionContext() {

